primitives/src/vid/mod.rs

```diff
@@ -0,0 +1,77 @@
+//! Trait and implementation for a Verifiable Information Retrieval (VID).
+/// See <https://arxiv.org/abs/2111.12323> section 1.3--1.4 for intro to VID semantics.
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use ark_std::{error::Error, fmt::Debug, string::String, vec::Vec};
+use displaydoc::Display;
+
+pub mod advz;
+
+/// The error type for `VidScheme` methods.
+#[derive(Display, Debug)]
+pub enum VidError {
+    /// invalid args: {0}
+    Argument(String),
+    /// internal error: {0}
+    Internal(anyhow::Error),
+}
+
+impl Error for VidError {}
+
+/// Convenience wrapper to convert any error into a [`VidError`].
+///
+/// Private fn so as not to expose error conversion API outside this crate
+/// as per [stackoverflow](https://stackoverflow.com/a/70057677).
+///
+/// # No-std support
+/// `no_std` mode requires `.map_err(vid)` to convert from a non-`anyhow` error
+/// as per [`anyhow` docs](https://docs.rs/anyhow/latest/anyhow/index.html#no-std-support),
+fn vid<E>(e: E) -> VidError
+where
+    E: ark_std::fmt::Display + Debug + Send + Sync + 'static,
+{
+    VidError::Internal(anyhow::anyhow!(e))
+}
+
+/// Convenience [`Result`] wrapper for [`VidError`].
+pub type VidResult<T> = Result<T, VidError>;
+
+/// VID: Verifiable Information Dispersal
+pub trait VidScheme {
```

Comment:
I wonder if we should add two associated constants as the config for an instantiation of this VID scheme, namely `payload_chunk_size` and `num_storage_nodes`?

Reply:
I can imagine applications where the user might want to change these values dynamically at runtime, in which case they cannot be compile-time constants. (eg. number of storage nodes might change with time, and `payload_chunk_size` is often a function of `num_storage_nodes`.)

In any case, this topic is outside the scope of this PR. We could post a separate issue if we don't want to forget about it in the future.